### ISSUE TITLE
Ecoflow River Pro integration add step runs forever

### ISSUE THREAD

**Issue by merlin-zaraza** [vwt12eh8/hassio-ecoflow#64]
I'm trying to add integration for my Ecoflow River Pro, but it's run forever on IP step.

<img width="413" alt="Screenshot 2022-12-05 at 18 51 42" src="https://user-images.githubusercontent.com/5223485/205695124-e04b00ca-40ed-481b-81a9-ecdc67c70972.png">

Home Assistant 2022.11.5
HACS 1.28.4
Ecoflow River Pro Firmware V1.2.5.27/V3.0.1.11(wi-fi)

**Comment by Ne0-Hack3r** [comment]
The v3.0.1.11 update for WiFi seems to be what breaks the integration (closes port 8055). That firmware only applies to DP units with certain WiFi chips (not all units have the same WiFi chip). But may only be a matter of time before an update closes the port on units as yet unaffected...

**Comment by merlin-zaraza** [comment]
@Ne0-Hack3r Got it. Somehow Ecoflow mobile app works, maybe you can figure out how? Thanks

**Comment by Ne0-Hack3r** [comment]
> @Ne0-Hack3r Got it. Somehow Ecoflow mobile app works, maybe you can figure out how? Thanks

The app works locally over Bluetooth or 'direct wireless' (ad Hoc), depending on the device, so it does not appear to use the local TCP connection that was open on some devices (at least not anymore).

Over WiFi the device communicates with a cloud server and will retrieve or publish MQTT messages on a set of defined topics. The app, in turn, communicates with that same MQTT server and, using MQTT messaging, exchanges status, settings, commands, etc.

Others, including myself, have begun tapping into the MQTT data as an alternative method of integration (though less elegant and more labor intense than using this integration to be sure).

Maybe EF will eventually have an advanced setting that allows us to enable port 8055 for specific IPs (like the IP of the HA server) so we can have private local integration without the security concern of having a wide open port...

**Comment by garithd** [comment]
Anyone care to share their HA MQTT "workaround" config?

**Comment by mmiller7** [comment]
I'm experiencing the same thing with a Delta Max 2000.  Add is hanging forever spinning.

This is the first I've seen anything about MQTT, I would be interested in any info people have on that...suppose I could MITM and Wireshark it and speculate whether it's using SSL or not but if someone has a guide that would be even easier.

**Comment by Ne0-Hack3r** [comment]
Check out this thread if you want to attempt MQTT integration... there are a number of steps and pieces involved... 

https://github.com/v1ckxy/ecoflow-withoutflow/issues/1

**Comment by mmiller7** [comment]
> Check out this thread if you want to attempt MQTT integration... there are a number of steps and pieces involved...
> 
> [v1ckxy/ecoflow-withoutflow#1](https://github.com/v1ckxy/ecoflow-withoutflow/issues/1)

Thanks - I'm making a bit of progress...at least got [a script written to reliably fetch credentials for MQTT](https://github.com/mmiller7/ecoflow-withoutflow/blob/main/ecoflow_get_mqtt_login.sh).  Now just have to work out what data exists, and what data I care about.

Most people seem to not have much info about sending control commands yet, bit of a bummer.  Would be cool to like have it command the Ecoflow to begin charging when there's a storm approaching so it can be stored at like 75% but topped up ahead of bad weather, and then brought back down again (running the inverter or something to bring the % back down faster)

**Comment by mmiller7** [comment]
> Check out this thread if you want to attempt MQTT integration... there are a number of steps and pieces involved...
> 
> [v1ckxy/ecoflow-withoutflow#1](https://github.com/v1ckxy/ecoflow-withoutflow/issues/1)

Seems like Home Assistant only allows one MQTT connection...so that may be a dead-end since I already use MQTT for things on my own broker.
![image](https://user-images.githubusercontent.com/606945/207741723-853a642f-d85a-44d8-be2d-b6a18421c745.png)

**Comment by Ne0-Hack3r** [comment]
> Most people seem to not have much info about sending control commands yet, bit of a bummer. Would be cool to like have it command the Ecoflow to begin charging when there's a storm approaching so it can be stored at like 75% but topped up ahead of bad weather, and then brought back down again (running the inverter or something to bring the % back down faster)

There are lots of clues in that thread... though I don't think anyone has published an actual list... If you setup MQTT Explorer and watch the ../set topic for the given device while changing settings in the app you can start collecting the payloads required to make changes using HA...

**Comment by Ne0-Hack3r** [comment]
> Seems like Home Assistant only allows one MQTT connection...

I have not gone down this road, yet, but it is my understanding that the Mosquitto Broker can support multiple bridges... I imagine you'd want to be careful about mapping to unique local topics to be sure messages to/from the disparate MQTT servers are routed properly...

This thread might point you in the right direction:
https://stackoverflow.com/questions/48982780/bridging-an-mqtt-broker-to-2-remote-mqtt-brokers-simultaneously

**Comment by mmiller7** [comment]
I think I finally got the MQTT stuff working after a solid 3 days of digging deep into it.  And I am reasonably happy with the results, aside from the fact it uses TheCloud and not local control (this will be my first ever not-local-control IoT device).

I've tried to document my efforts on my own GitHub: https://github.com/mmiller7/ecoflow-withoutflow/tree/main/cloud-mqtt

But I'm overall pleased:
![image](https://user-images.githubusercontent.com/606945/208534585-0dc95a47-9796-40f8-8f50-b1961713b207.png)

**Comment by Ne0-Hack3r** [comment]
Nice... I see you're making good use of the multi-entity row module as well (I just discovered it recently myself)...

**Comment by mmiller7** [comment]
> Nice... I see you're making good use of the multi-entity row module as well (I just discovered it recently myself)...

Yes, I have also learned about SO many new template integration functions this past 3 days that I never knew previously existed!

Also learned on the "Glance" card you can put other stuff like "name" or "icon" in the YAML just like the Entities card, it just doesn't have a UI editor to populate the fields.  I have much more verbose names "under the hood" but simple names on the display.

**Comment by Ne0-Hack3r** [comment]
> Also learned on the "Glance" card you can put other stuff like "name" or "icon" in the YAML just like the Entities card, it just doesn't have a UI editor to populate the fields. I have much more verbose names "under the hood" but simple names on the display.

Cool... I'll have to look into that one... Yes, this project has immersed me in JSON, YAML, and much more of what can be done with HA... a bit of a learning curve at first but I'm quickly becoming much more comfortable and advanced...

**Comment by skpow55** [comment]
Excellent work on this!

**Comment by STaN33130** [comment]
> I think I finally got the MQTT stuff working after a solid 3 days of digging deep into it. And I am reasonably happy with the results, aside from the fact it uses TheCloud and not local control (this will be my first ever not-local-control IoT device).
> 
> I've tried to document my efforts on my own GitHub: https://github.com/mmiller7/ecoflow-withoutflow/tree/main/cloud-mqtt
> 
> But I'm overall pleased: ![image](https://user-images.githubusercontent.com/606945/208534585-0dc95a47-9796-40f8-8f50-b1961713b207.png)

I just install it a big work ! I have a question, is it possible to control the switches like AC output on your configuration or no ?

**Comment by mmiller7** [comment]
> 
> I just install it a big work ! I have a question, is it possible to control the switches like AC output on your configuration or no ?

Yes, that seems to work.  I had to use the advice in this thread to use MQTT explorer to capture the commands my Delta Max uses for the on/off from the app so I could re-play them correctly from Home Assistant as a template-switch action.

All the sliders, switches, and option-selects in the screenshot I was able to get working for my Delta Max by carefully capturing each app-action in MQTT explorer and writing a template/script to replay the right commands to the same MQTT topic.

I don't know if it may change what is sent for each model battery

**Comment by STaN33130** [comment]
i probably have made a mistake because when i click on a switch in HA, nothing happens in EF app. But when i click on a switch in EF app, no problem to see it in HA. I think there is a problem with the MQTT publish (but it's probably my fault!... I'm a MQTT noob)

**Comment by mmiller7** [comment]
> i probably have made a mistake because when i click on a switch in HA, nothing happens in EF app. But when i click on a switch in EF app, no problem to see it in HA. I think there is a problem with the MQTT publish (but it's probably my fault!... I'm a MQTT noob)

If you have a different model than the delta max it might use a different command.  May have to figure out what it sends from the app and modify the switch sensors

**Comment by STaN33130** [comment]
you're right. I have a Delta 2 with extra battery and a River 2 max. Most of the sensors are different compare to your battery. Thank you for your help

**Comment by Ne0-Hack3r** [comment]
> you're right. I have a Delta 2 with extra battery and a River 2 max. Most of the sensors are different compare to your battery. Thank you for your help

Delta 2 operates quite a bit different than Delta Pro / Delta Max - I have a full package built for Delta 2 and I ended up using a "workaround" where anytime any sensor is updated, and D2 publishes that change, one automation publishes a 'request for all states' to ../get and another automation re-publishes the resulting payload with 'state of everyting' from ../get_reply to ../status - thus I built most of my sensors around data in ../status -- still feel a bit kludgy but when I tried building all sensors around the data periodically published by D2 I ended up with a lot of sensors that sat for long periods of time in "unknown" as the D2 only appears to publish the state of certain sensor when/if they change... In contrast, Delta Pro, for example, publishes the state of all sensors for a given module (EMS, BMS, Inverter, MPPT, etc.) frequently (every 1-5 seconds) so as long as you build sensors with logic to capture the desired data "if" it exists in the payload everything populates quickly and stays up to date... 

It looks like, in the case of all EF devices I've looked at thus far, the native app will query for 'state of everything', populate the data into the app and then watch for changes as published... This is a bit tricky for HA and could require a lot more code/logic to implement... So, unless/until I or someone else discover a better way to populate HA sensors given this behavior I'm using this workaround...

**Comment by Ne0-Hack3r** [comment]
@STaN33130 

Have a look at:
https://github.com/v1ckxy/ecoflow-withoutflow/issues/1
https://github.com/v1ckxy/ecoflow-withoutflow/issues/6

**Comment by STaN33130** [comment]
Cool thank you for your help !
